### DISCUSSION TITLE
`markdown-enable-math` interferes with `flymake-goto-next-error` in an `eglot` managed buffer

### DISCUSSION THREAD

**Discussion by ywwry66** [joaotavora/eglot#1394] [Q&A] [answered]
Hi there,

Thanks for working on this great project!

I recently found a wierd issue when using `eglot` with `markdown-enable-math` set to `t`. Specifically, the error messages in the echo area produced by `flymake-goto-next-error` only flash for a second and disappear.

The following example is reproducible on my local environment (macOS Sonoma, latest stable `Emacs` and `R`, `languageserver` and `lintr` `R`-packages, `markdown-mode` from `melpa`, built-in `eglot` and `flymake`):
1. `emacs -Q`
2. Evaluate
```elisp
(require 'package)
(package-initialize)
(setq-default markdown-enable-math t)
```
3. Open `R` script containing
```r
## test
A <- 1
```
4. Run `eglot`
5. Run `flymake-goto-next-error`

I don't know whether this issue is caused by `eglot`, `flymake`, or `markdown-mode`. But I think you are the most knowledgeable in this regard and thus raised my question here.

**Comment by joaotavora** [comment, marked as answer]
I don't have the R tool chain, sorry. Maybe you'll have luck with someone
else here. Maybe you can add debug-on-entry to the message function to know
who is calling it to clear the echo area. But it's hard to get it to not
interfere with normal Emacs usage to much.

**Reply by ywwry66** [reply]
Thanks for the tips. I was able to figure out the reason:

Running `flymake-goto-next-error` produces the following call sequence `eglot-hover-eldoc-function` -> `eglot--hover-info` -> `eglot--format-markup` -> `gfm-view-mode` -> `gfm-mode` -> `markdown-mode`. If `markdown-enable-math` is set, `markdown-toggle-math` is called, which messages "markdown-mode math support enabled".

What do you think is the proper way of fixing this?

**Reply by joaotavora** [reply]
> What do you think is the proper way of fixing this?

I don't know honestly, but something like having Eglot inhibit messages inside the callees of eglot--format-markup, which is possible at least in Emacs 29.1 (maybe let-binding `set-message-functions`, which see).

**Reply by joaotavora** [reply]
Another equally good way, perhaps even more idiomatic, is for `markdown-toggle-math` to _not_ call `message` if it' snot being invoked interactively.

**Reply by ywwry66** [reply]
> I don't know honestly, but something like having Eglot inhibit messages inside the callees of eglot--format-markup, which is possible at least in Emacs 29.1 (maybe let-binding set-message-functions, which see).

It seems the inhibit message mechamism alway clears the echo area so that does not help. But I was able to use `set-message-functions` to fixed problem. The idea is to record what was already on the echo area and always output that message thereafter.
```elisp
  (defun existing-message-advice (fun &rest args)
    (let* ((existing-message (current-message))
	   (set-message-functions (list (lambda (msg) existing-message))))
      (apply fun args)))
  (advice-add 'eglot--format-markup :around #'existing-message-advice)
```
Of course this line needs to be removed: https://github.com/joaotavora/eglot/blob/db91d58374627a195b731a61bead9b4f84a7e4bc/eglot.el#L1812
> Another equally good way, perhaps even more idiomatic, is for markdown-toggle-math to not call message if it' snot being invoked interactively.

You are right this is actually better. I will open a pull request upstream.

**Reply by ywwry66** [reply]
Fixed by upstream.
